### PR TITLE
Let StorageDoubleMap use borrowed key types

### LINKED ISSUE #1566: use `AsRef<KeyType>` for Contract Storage
keys should also be able to take AsRef<KeyType> to ensure Vec<u8>s can be passed as &[u8]

https://github.com/paritytech/substrate/blob/bbe18168174cadaab8c693dd1f0812a883d150e4/srml/contract/src/lib.rs#L353-L361

### PR DESCRIPTION
**Pull request by holygits**
Closes #1566

### CODE REVIEW

File: srml/support/src/double_map.rs [hunk 1]
```diff
@@ -38,42 +39,78 @@ pub trait StorageDoubleMap {
...
 		unhashed::put(&Self::full_key(k1, k2)[..], &val);
 	}
 
 	/// Remove an entry from this map.
-	fn remove(k1: Self::Key1, k2: Self::Key2) {
+	fn remove<Q, R>(k1: &Q, k2: &R)
+	where
+		Self::Key1: Borrow<Q>,
+		Self::Key2: Borrow<R>,
+		Q: Codec,
```
Comment (gavofyork) [inline srml/support/src/double_map.rs]: why do `Q` and `R` need to be `Codec` if `Key1`/`Key2` can be borrowed from them anyway (and are already `Codec`)?

Comment (holygits) [reply srml/support/src/double_map.rs]: It tells the compiler that the borrowed key types (`Q` and `R`) implement Codec as well.
I followed the pattern used by `std::collections::HashMap` .e.g
```
    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
        where K: Borrow<Q>,
              Q: Hash + Eq
```
https://github.com/rust-lang/rust/blob/b244f61b77c42d7be695afd7901ee4418559e518/src/libstd/collections/hash/map.rs#L1300-L1302

Comment (gavofyork) [reply srml/support/src/double_map.rs]: Guess it's just something that `full_key` forces you into.

**Comment by gavofyork** [comment]
@holygits just waiting for it to be resolved...

File: srml/contract/src/account_db.rs [hunk 2]
```diff
@@ -54,7 +54,7 @@ pub trait AccountDb<T: Trait> {
 pub struct DirectAccountDb;
 impl<T: Trait> AccountDb<T> for DirectAccountDb {
 	fn get_storage(&self, account: &T::AccountId, location: &[u8]) -> Option<Vec<u8>> {
-		<StorageOf<T>>::get(account.clone(), location.to_vec())
+		<StorageOf<T>>::get(account, &location.to_vec())
```
Comment (gavofyork) [inline srml/contract/src/account_db.rs]:
```suggestion
		<StorageOf<T>>::get(account, location)
```

File: srml/contract/src/tests.rs [hunk 3]
```diff
@@ -224,13 +224,13 @@ fn account_removal_removes_storage() {
 			{
 				Balances::set_free_balance(&1, 110);
 				Balances::increase_total_stake_by(110);
-				<StorageOf<Test>>::insert(1, b"foo".to_vec(), b"1".to_vec());
-				<StorageOf<Test>>::insert(1, b"bar".to_vec(), b"2".to_vec());
+				<StorageOf<Test>>::insert(&1, &b"foo".to_vec(), b"1".to_vec());
```
Comment (gavofyork) [inline srml/contract/src/tests.rs]:
```suggestion
				<StorageOf<Test>>::insert(&1, b"foo", b"1".to_vec());
```

File: srml/contract/src/tests.rs [hunk 4]
```diff
@@ -224,13 +224,13 @@ fn account_removal_removes_storage() {
...
+				<StorageOf<Test>>::insert(&1, &b"bar".to_vec(), b"2".to_vec());
```
Comment (gavofyork) [inline srml/contract/src/tests.rs]: ditto...

**Comment by holygits** [comment]
Hmm, I get lots of errors like the following:
```
the trait `_IMPL_DECODE_FOR_RawEvent::_parity_codec::Decode` is not implemented for `[u8]`
```

I think the encode trait for this would just call `.to_vec()` anyway
https://github.com/paritytech/parity-codec/blob/50ea4fc9c74a426720ff62ed166c9ced485cac89/src/codec.rs#L135-L140

I'm not sure what the best path forward is here

**Comment by gavofyork** [comment]
Ok; probably not sensible to try to solve in this PR.

**Pull request merged**